### PR TITLE
.NET smoke Test Sample

### PR DESCRIPTION
**Pull request by JonathanCrd**
Performs functionalities with Key Vault, Identity, Event Hubs and Blob Storage Track 2 SDKs.

### CODE REVIEW

**Review by danieljurek — COMMENTED** [review]
This is a great first cut! This review covers stylistic and a few things dealing with the general logic of the app without diving too far into program design (except the use of `static`). To that end we'll probably have another review where, after we get the style discussions out of the way, we can dig into the exciting details of Object Oriented design.

File: samples/SmokeTest/SmokeTest/BlobStorage.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+         * The connection string is retreived from an envirmonet variable.
+         * The container name for this sample is 'mycontainer', and the Blob name 'SmokeTestBlob'
+         */
+
+        private static BlobServiceClient service = new BlobServiceClient(Environment.GetEnvironmentVariable("Blob_ConnectionString"));
+        private static BlockBlobClient blob = service.GetBlobContainerClient("mycontainer").GetBlockBlobClient("SmokeTestBlob");
+
+        public static async Task performFunctionalities()
+        {
+            Console.WriteLine("\n---------------------------------");
```
[the remark this answers is not in the document]
Comment (danieljurek) [reply samples/SmokeTest/SmokeTest/BlobStorage.cs]: Work on this is lower priority given the larger goals of the project. It's worth noting that a unit test framework like xUnit or NUnit can handle this work with little overhead.

File: samples/SmokeTest/SmokeTest/Program.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+namespace SmokeTest
+{
+
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            Console.WriteLine("SMOKE TEST FOR TRACK 2 LIBRARIES");
+            await KeyVaultTest.performFunctionalities();
+            await BlobStorage.performFunctionalities();
```
Comment (danieljurek) [inline samples/SmokeTest/SmokeTest/Program.cs]: `KeyVaultTest` is a great name. We should standardize the names of these classes so it's clear that their goals are similar. For example `BlobStorage` should be `BlobStorageTest` and `EventHubs` should be `EventHubsTest`

File: samples/SmokeTest/SmokeTest/BlobStorageTest.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
+﻿using Azure.Storage.Blobs;
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices.ComTypes;
+using System.Threading.Tasks;
+
+namespace SmokeTest
+{
+
```
Comment (weshaggard) [inline samples/SmokeTest/SmokeTest/BlobStorageTest.cs]: nit: remove the blank lines.

File: samples/SmokeTest/SmokeTest/TestBase.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+namespace SmokeTest
+{
+    abstract class TestBase
+    {
+        /// <summary>
+        /// High order function to handle the exceptions of each test.
+        /// </summary>
+        /// <param name="testAction"></param>
+        /// <returns></returns>
+        protected async Task<bool> ExcecuteTest(Func<Task> testAction)
```
Comment (weshaggard) [inline samples/SmokeTest/SmokeTest/TestBase.cs]: typo: `ExcecuteTest` -> `ExecuteTest`

File: samples/SmokeTest/SmokeTest/BlobStorageTest.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            Console.WriteLine("Functionalities to test: 2:");
+            Console.WriteLine("1.- Upload Blob Block");
+            Console.WriteLine("2.- Delete that Blob Block" + '\n');
+
+            Console.Write("Uploading blob... ");
+            var result1 = await ExcecuteTest(UploadBlob);
+            if (!result1)
+            {
+                //If this test failed, then the othe one is going to fail too.
+                Console.WriteLine("Cannot delete the Blob.");
```
Comment (weshaggard) [inline samples/SmokeTest/SmokeTest/BlobStorageTest.cs]: I'd suggest you passing in the message to ExecuteTest method and show it in the exception handler in that method.

**Comment by weshaggard** [comment]
@JonathanCrd I'd suggest having a look at https://github.com/Azure/azure-sdk-for-net/pull/6678 which is doing something very similar to what you are doing and so the pattern might be interesting here as well.

**Comment by pakrym** [comment]
Is the intent here to have a set of samples or to have a set of tests?

**Comment by weshaggard** [comment]
>Is the intent here to have a set of samples or to have a set of tests?

The intent is actually to run a set of samples for each of the services in a single application. For now we are just writing samples independently but the goal is to eventually pull in all the samples from the library directory into this single application.

**Comment by pakrym** [comment]
The way it's written right now looks like an attempt to reinvent testing framework.

**Comment by weshaggard** [comment]
>The way it's written right now looks like an attempt to reinvent testing framework.

I agree and we should definitely avoid duplicating a testing framework. I kind of like what @jsquire did in https://github.com/Azure/azure-sdk-for-net/pull/6678 which is why I pointed to that as an example.

**Comment by pakrym** [comment]
I wonder if we can build on that idea but avoid ISample all together. Like, have every file be a self-contained copy-pastable app with Program.Main but also have a common runner that is able to run multiple samples.

**Comment by jsquire** [comment]
> I wonder if we can build on that idea but avoid ISample all together. Like, have every file be a self-contained copy-pastable app with Program.Main but also have a common runner that is able to run multiple samples.

I used an interface just to enforce structure;  in the context of Event Hubs, I wanted to ensure that every sample had a consistent method to execute which accepted a common set of connection parameters (to avoid ambient context and the need to set environment variables; I wanted to enable the F5 experience in Visual Studio.)     I could have made implicit assumptions about the structure to avoid the interface, but preferred to make it explicit in case there were community contributions.

With respect to a generic approach that would span client libraries, I'd expect that to be tough without ending up with the need for ambient environment requirements again.  In my opinion, having an approachable "it just works" experience for being able to launch the debugger is the more desirable side of that trade off.

**Comment by weshaggard** [comment]
I agree it is something to build on and I'm fine making them standalone but we still need a way to aggregate them together.

**Comment by pakrym** [comment]
> With respect to a generic approach that would span client libraries, I'd expect that to be tough without ending up with the need for ambient environment requirements again. In my opinion, having an approachable "it just works" experience for being able to launch the debugger is the more desirable side of that trade off.

I wonder if we can use System.CommandLine (https://github.com/dotnet/command-line-api) inside sample apps themselves making them explicit and directly callable.

File: samples/SmokeTest/SmokeTest/Program.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace SmokeTest
+{
+    class Program
+    {
+        static async Task<int> Main(string[] args)
+        {                        
+            Console.WriteLine("SMOKE TEST FOR TRACK 2 LIBRARIES");
+
+            var keyVault = new KeyVaultTest("SmokeTestSecret", "smokeTestValue", Environment.GetEnvironmentVariable("DIR_TENANT_ID"), Environment.GetEnvironmentVariable("APP_CLIENT_ID"), Environment.GetEnvironmentVariable("CLIENT_SECRET"), Environment.GetEnvironmentVariable("KEY_VAULT_URI"));
```
Comment (weshaggard) [inline samples/SmokeTest/SmokeTest/Program.cs]: We should turn these into static method calls and move the configuration into the individual sample file.

File: samples/SmokeTest/SmokeTest/BlobStorageTest.cs (hunk 5 above)
Comment (JonathanCrd) [reply samples/SmokeTest/SmokeTest/BlobStorageTest.cs]: Resolving this review since the high order function is not going to be used anymore.

**Review by danieljurek — APPROVED** [review]

**Pull request merged**